web/src/compose_tooltips.ts

```diff
@@ -14,10 +14,75 @@ import {pick_empty_narrow_banner} from "./narrow_banner.ts";
 import * as narrow_state from "./narrow_state.ts";
 import * as popover_menus from "./popover_menus.ts";
 import {realm} from "./state_data.ts";
-import {EXTRA_LONG_HOVER_DELAY, INSTANT_HOVER_DELAY, LONG_HOVER_DELAY} from "./tippyjs.ts";
+import {
+    EXTRA_LONG_HOVER_DELAY,
+    INSTANT_HOVER_DELAY,
+    LONG_HOVER_DELAY,
+    SINGLETON_INSTANT_HOVER_DELAY,
+    SINGLETON_LONG_HOVER_DELAY,
+    get_tooltip_content,
+} from "./tippyjs.ts";
 import {parse_html} from "./ui_util.ts";
 import {user_settings} from "./user_settings.ts";
 
+type SingletonContext = "compose" | `edit_message:${string}`;
+type SingletonTooltips = {
+    tooltip_instances: tippy.Instance[] | null;
+    singleton_instance: tippy.CreateSingletonInstance | null;
+};
+
+const compose_button_singleton_context_map = new Map<SingletonContext, SingletonTooltips>();
+
+export function clean_up_existing_instances(context: SingletonContext): void {
```

Comment:
Let's call this `clean_up_compose_singleton_tooltip` instead, and include a comment as to why we need this function. You can include a reference to http://atomiks.github.io/tippyjs/v6/addons/#destroy.